### PR TITLE
[NVTPX] Copy kernel arguments as byte array

### LINKED ISSUE #53710: CUDA: unions in kernel arguments not copied completely if member contains padding
**Summary**

I have a CUDA kernel that takes a `union` as one of its parameters. One of the `union`'s member types is a `struct` (called `Ptrs` in my minimal working example) that contains some padding due to the memory alignment requirements of its own members. I initialize the `union` as its other member type (`Data` in the MWE) and pass it to the kernel launch. Inside the kernel however, some of the contents of `Data` appear to be zeroed out -- notably those corresponding to the padding region of `Ptrs`. It seems like Clang looks at the padding inside one of the `union`'s member types and assumes that all the members have holes in the same places.

When compiling the CUDA kernel, Clang should make no assumption about which one of the member types is currently stored in a `union` that it gets as an argument, but it clearly does and it's the wrong assumption in this case. We initially found this bug in conjunction with [thrust](https://github.com/nvidia/thrust) and [mpark::variant](https://github.com/mpark/variant), but it can be observed with plain CUDA code too as in the MWE.

**Versions**

I reproduced this bug with multiple Clang versions between 7.0.0 and 14.0.0-rc1, as well as the current master branch. Nvidia's nvcc compiler does not exhibit such an issue.

Starting with https://github.com/llvm/llvm-project/commit/d0615a93bb6d7aedc43323dc8957fe57e86ed8ae, my MWE required a minor modification to make sure that the argument copy wasn't optimized out. This optimization is not generally possible (e.g. when using thrust), so we cannot rely on it accidentally fixing the real issue.

**Minimal working example**

Compile with `clang++ -O3 -std=c++14 --cuda-gpu-arch=sm_70 mwe.cu -o mwe.cu -lcudart`. The issue can also be observed without `-O3`. Add `-DFIX_IT` to get a struct without padding, in which case the issue does not occur.

```cuda
#include <algorithm>
#include <iostream>
#include <vector>

#include <cuda.h>

struct Ptrs
{
  char s1, s2, s3, s4, s5, s6, s7;
#ifdef FIX_IT
  char s8;
#endif
  void const * data;
};

template <class T, unsigned N>
struct Data
{
  T data[N];
};

template <class T, unsigned N>
union DataType
{
  Data<T,N> d;
  Ptrs p;
};

__global__ void transform(DataType<unsigned, 4> umap, unsigned n, unsigned * result)
{
  const unsigned i = threadIdx.x;

  if(n > 0xffff) {
    /* This condition is never true, but the next line ensures that `umap` is copied into writable memory.
     * Clang 13 (https://github.com/llvm/llvm-project/commit/d0615a93bb6d7aedc43323dc8957fe57e86ed8ae)
     * introduced an optimization that will otherwise mask the issue we are trying to demonstrate.
     */
    umap.d.data[i] = 0;
  }

  if(i < n)
  {
    result[i] = umap.d.data[i];
  }
}

bool test()
{
  std::vector<unsigned> map{23U, 0xffffffff, 42U, 13U};
  DataType<unsigned, 4> umap;
  Data<unsigned, 4> d;
  std::copy(map.begin(), map.end(), &d.data[0]);
  umap.d = d;

  const unsigned mod = map.size();

  std::vector<unsigned> h_values(16);

  unsigned * values = nullptr;
  cudaMalloc(&values, h_values.size() * sizeof(unsigned));
  transform<<<1, h_values.size()>>>(umap, h_values.size(), values);
  cudaMemcpy(h_values.data(), values, h_values.size() * sizeof(unsigned), cudaMemcpyDeviceToHost);
  cudaFree(values);

  bool good = true;
  for(int i = 0; i < mod; ++i)
  {
    std::cout << h_values[i];
    if (h_values[i] != map[i])
    {
      std::cout << " wrong";
      good = false;
    }
    std::cout << std::endl;
  }

  return good;
}

int main()
{
    bool good = test();
    return (good ? 0 : 1);
}
```

**Assembly**

To see what is happening, compile the MWE with and without `-DFIX_IT` and compare the NVPTX assembly: https://godbolt.org/z/W7PY9q3sj . You can see that in the former case, it generates `ld.param` and `st.local` instructions that cover the entire size of the `union`, while in the latter case it skips those bytes that are padding.

### PR DESCRIPTION
**Pull request by mkuron**
Ensures that struct padding is not skipped, as it may contain actual data if the struct is really a union.

The patch originated from a discussion on #53710, and @Artem-B suggested I post it as a pull request. This is my first LLVM code contribution, so I may need a bit of help in getting this into a shape acceptable to merge.

I updated the existing test cases that were failing after my change, which lead to some changes from `align 4` to `align 8` on the `store`. The result actually looks correct to me -- the corresponding argument and the corresponding `alloca` both have `align 8`, only the `store` previously used a narrower alignment than it could have.

Fixes #53710

### CODE REVIEW

**Review by Artem-B — COMMENTED** [review]
I'm apparently bad at clicking on  "Submit review" button. This comment's been sitting there unsubmitted for a while. Sorry about that.

File: llvm/lib/Target/NVPTX/NVPTXLowerArgs.cpp
Comment (mkuron) [reply llvm/lib/Target/NVPTX/NVPTXLowerArgs.cpp:644]: @Artem-B, here is one: https://godbolt.org/z/7EKzMEeK3. That pointer to a `{ [45 x i8] }` has 16-byte alignment, so should vectorize into a `{ [2 x <2 x i64>], i64, i32, i8 }`. You can see that _load-store-vectorizer_ runs after _nvptx-lower-args_, but does nothing. Debug output shows messages like `LSV: Access of ... in addrspace 5 with alignment 1 is misaligned, and therefore can't be vectorized.`, which looks like some alignment is not preserved. Once you find a fix for that, I'll remove the manual vectorization from my pull request.

Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXLowerArgs.cpp:644]: Interesting. I think the problem is that `SROA` just sticks 'align 1' on all 1-byte loads/stores, because it technically does not need any higher alignment, and LSV takes it at face value, even though the base pointer is aligned.

That looks like something we may want to fix, eventually, but it's probably going to take some time. 

Can we just use memcpy instead of load/store w/ a custom type? That should copy the data we need *and* convey proper alignment info to the code that does the copying.

Comment (mkuron) [reply llvm/lib/Target/NVPTX/NVPTXLowerArgs.cpp:644]: > Can we just use memcpy instead of load/store w/ a custom type?

Yes, we should be able to do that, I just couldn’t figure out how.

Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXLowerArgs.cpp:644]: Here's an example of creating an intrinsic call: https://github.com/llvm/llvm-project/blob/fc4b1a303b296d02f6243a083510c4ee7f290ab0/llvm/lib/CodeGen/IntrinsicLowering.cpp#L465
Memcpy should be similar, though it may need extra parameters to get the right intrinsic declaration as they are overloaded on pointers' AS.

Comment (mkuron) [reply llvm/lib/Target/NVPTX/NVPTXLowerArgs.cpp]: Thanks @Artem-B, that actually turned out to be quite easy. Latest commit switches to memcpy and updates tests accordingly.

File: llvm/test/tools/UpdateTestChecks/update_llc_test_checks/Inputs/nvptx-basic.ll.expected
Comment (Artem-B) [inline llvm/test/tools/UpdateTestChecks/update_llc_test_checks/Inputs/nvptx-basic.ll.expected:31]: Oh. This is nice! I have noticed these pointless re-loads from the local copy and had tracking them down on my todo list. Looks like your change takes care of this.

File: llvm/test/CodeGen/NVPTX/lower-args.ll
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/lower-args.ll:41]: Negative checks are tricky to get right. In general, they should be done in a separate run with only negative checks and the positive 'anchor` checks around the areas of interest.

As written, if a per-field copy happens before `ld.param.u64` the test will still pass.

Autogenerated LLC checks capturing complete set of instructions would work best here, though the check generator has issues handling tests like these that combine llc and opt runs.

**Review by Artem-B — APPROVED** [review]
LGTM, with a test nit.

File: llvm/test/CodeGen/NVPTX/lower-args.ll
Comment (mkuron) [reply llvm/test/CodeGen/NVPTX/lower-args.ll:41]: I guess I could just remove the negative checks here. The struct has size 64, so if there is an `ld.param.u64`, there can't be any of the others.

Comment (Artem-B) [reply llvm/test/CodeGen/NVPTX/lower-args.ll:41]: That would still check for "one thing looks right" while it would ignore all the things that may have gone wrong (and  catching the things that are wrong is arguably as important as verifying the things that we expect to happen).

E.g. that redundant re-load from the local memory, now eliminated by your patch, would be ignored, if it were still around. So, your test may have checked for ld.param.u64 while we'd be copying into that local array, but would not notice if we'd re-load from that array field-by-field.

This case is trivial, so I'd just run the update_llc_test_checks.py on the test and use the checks it produced, possibly restoring OPT checks if they get nuked in the process. One of the days we should teach `update_test_checks.py` and `update_llc_test_checks.py` to coexist in the same test file.

Comment (mkuron) [reply llvm/test/CodeGen/NVPTX/lower-args.ll]: Latest commit adds a PTX check for the entire function.

**Review by Artem-B — APPROVED** [review]

**Comment by mkuron** [comment]
@Artem-B, do we need another approval? Who can merge?

**Comment by Artem-B** [comment]
You are good to go. If you do not have commit access I can merge the patch on your behalf.

**Comment by mkuron** [comment]
@Artem-B, please do merge it.

**Pull request merged**
